KLR/HLR/AST.lean

```diff
@@ -0,0 +1,278 @@
+  statements : List Statement
+deriving Inhabited, Repr, Nonempty
+
+-- An HLR program
+structure Program where
+  functions : List Function
+deriving Inhabited, Repr, Nonempty
+
+-- Returns the list of variables that this operator immediately depends on.
+def dependencies : Operator → List Var
+  | .arg _ => []
+  | .binaryOp _ a b => [a, b]
+  | .unaryOp _ a => [a]
+  | .reductionOp _ a b _ => [a, b]
+  | .batchMatmul a b => [a, b]
+  | .arange .. => []
+  | .concat tensors _ => tensors
+  | .select cond a b => [cond, a, b]
+  | .full .. => []
+  | .transpose a _ => [a]
+  | .split_with_sizes a _ => [a]
+  | .reshape a _ => [a]
+  | .broadcast a .. => [a]
+  | .const .. => []
+  | .gather a i .. => [a, i]
+  | .slice a .. => [a]
+  | .call _ inputs => inputs
+
+-- Returns the list of all variables in this function.
+def vars (f : Function) : List Var :=
+  f.statements.filterMap (fun stmt =>
+    match stmt with
+    | .assign dest .. => .some dest
+    | _ => .none)
+
+-- Finds the operator that assigns to a variable in the function.
+def findVar (f : Function) (v : Var) : Option Operator :=
+  f.statements.findSome? (fun stmt =>
+    match stmt with
+    | .assign dest op _ => if dest == v then .some op else .none
+    | _ => .none)
+
+-- Returns the list of all variables that this function depends on transitively.
+def transitiveDependencies (f : Function) (v : Var) : List Var := Id.run do
+  let mut deps := [v]
+  repeat
+    let newDeps := deps.flatMap (fun v => match findVar f v with
+      | .some x => dependencies x
+      | .none => [])
+    let filtered := newDeps.filter (fun s => ! deps.contains s)
+    if filtered.isEmpty then
+      break
+    deps := deps ++ filtered
+  return deps
+
+-- TODO: move these toString instances to the TensorLib repo
+instance : ToString TensorLib.Slice where
+  toString s :=
+    let {start, stop, step, ..} := s
+    let start := start.map toString |>.getD ""
+    let stop := stop.map toString |>.getD ""
+    let step := step.map toString |>.getD ""
+    s!"{start}:{stop}:{step}"
+
+instance : ToString TensorLib.Shape where
+  toString s :=
+    s.val.map toString |> "x".intercalate |> fun x => s!"[{x}]"
+
+instance : ToString TensorLib.Dtype where
+  toString
+    | .bool => "bool"
+    | .int8 => "i8"
+    | .int16 => "i16"
+    | .int32 => "i32"
+    | .int64 => "i64"
+    | .uint8 => "u8"
+    | .uint16 => "u16"
+    | .uint32 => "u32"
+    | .uint64 => "u64"
+    | .float32 => "f32"
+    | .float64 => "f64"
+
+instance : ToString BinaryOp where
+  toString
+    | .mul => "mul"
+    | .max => "max"
+    | .sub => "sub"
+    | .add => "add"
+    | .div => "div"
+    | .cmp => "cmp"
+    | .and => "and"
+
+instance : ToString UnaryOp where
+  toString
+    | .exp => "exp"
+    | .sqrt => "sqrt"
+    | .neg => "neg"
+    | .convert dtype => s!"convert_{dtype}"
+
+instance : ToString TensorTy where
+  toString (t : TensorTy) : String :=
+    s!"{t.shape}{t.dtype}"
+
+instance : ToString Operator where
```

Comment:
I recommend defining a function of type `a -> Format` and calling `pretty` on that to get a string. Otherwise formatting of this will be very hard to read. If you're happy with a 1-line version of this it's fine, but it may be harder to write a formatter after the fact.